### PR TITLE
Introducing loop support for NME

### PR DESCRIPTION
**Pull request by deltakosh**

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Review by sebavan — APPROVED** [review]

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15599/merge

Test environment: 
https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15599/merge/index.html

To test a playground add it to the URL, for example:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15599/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15599/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15599/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15599/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15599/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15599/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15599/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15599/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15599/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for WebGL 1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://snapshots-cvgtc2eugrd3cgfd.z01.azurefd.net/refs/pull/15599/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

File: packages/dev/core/src/Materials/Node/Blocks/loopBlock.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
+import { NodeMaterialBlock } from "../nodeMaterialBlock";
+import { NodeMaterialBlockConnectionPointTypes } from "../Enums/nodeMaterialBlockConnectionPointTypes";
+import type { NodeMaterialBuildState } from "../nodeMaterialBuildState";
+import { NodeMaterialConnectionPointDirection, type NodeMaterialConnectionPoint } from "../nodeMaterialBlockConnectionPoint";
```
Comment (Popov72) [inline packages/dev/core/src/Materials/Node/Blocks/loopBlock.ts:4]: `type` should not be used inside an `import` (that would be cool if we could tell VSC not to do that...). There are other occurrences in other files.

File: packages/dev/core/src/Materials/Node/Blocks/storageReadBlock.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+     * Gets the value component
+     */
+    public get value(): NodeMaterialConnectionPoint {
+        return this._outputs[0];
+    }
+
+    protected override _buildBlock(state: NodeMaterialBuildState) {
+        super._buildBlock(state);
+
+        const value = this._outputs[0];
```
Comment (Popov72) [inline packages/dev/core/src/Materials/Node/Blocks/storageReadBlock.ts:57]: nit: `const value = this.value;` ?

File: packages/dev/core/src/Materials/Node/Blocks/storageWriteBlock.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        }
+        const loopBlock = this.loopID.connectedPoint!.ownerBlock as LoopBlock;
+
+        return loopBlock.output.isConnectedInFragmentShader;
+    }
+
+    protected override _buildBlock(state: NodeMaterialBuildState) {
+        super._buildBlock(state);
+
+        const value = this._inputs[1];
```
Comment (Popov72) [inline packages/dev/core/src/Materials/Node/Blocks/storageWriteBlock.ts:69]: nit: `const value = this.value;`?

File: packages/dev/core/src/Materials/Node/Blocks/storageReadBlock.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import type { NodeMaterialBuildState } from "../nodeMaterialBuildState";
+import { NodeMaterialConnectionPointDirection, type NodeMaterialConnectionPoint } from "../nodeMaterialBlockConnectionPoint";
+import { NodeMaterialBlockTargets } from "../Enums/nodeMaterialBlockTargets";
+import { RegisterClass } from "../../../Misc/typeStore";
+import { LoopBlock } from "./loopBlock";
+import { NodeMaterialConnectionPointCustomObject } from "../nodeMaterialConnectionPointCustomObject";
+/**
+ * Block used to read from a variable within a loop
+ */
+export class StorageReadBlock extends NodeMaterialBlock {
```
Comment (Popov72) [inline packages/dev/core/src/Materials/Node/Blocks/storageReadBlock.ts:12]: `StorageReadBlock` and `StorageWriteBlock` can only be used in connection with a `Loop` block (the `loopID` input must be connected to a `Loop` block), so maybe add "Loop" somewhere in the class name to make it clear these classes work together?

**Review by Popov72 — COMMENTED** [review]
I won't lie to you, I didn't understand everything: I trust you that it's working as expected! :)

File: packages/dev/core/src/Materials/Node/Blocks/storageReadBlock.ts (hunk 4 above)
Comment (deltakosh) [reply packages/dev/core/src/Materials/Node/Blocks/storageReadBlock.ts:12]: Well I'm not a big fan of over indexing with a longer name. As the ports are mandatory and linked to a loop, I feel it is enough

**Pull request merged**